File: swimming-contest-software/Networking/src/main/java/services/SwimmingRaceServicesProxy.java
package services;

import domain.dtos.AdminDTO;
import domain.dtos.RaceDTO;
import domain.dtos.RaceDetailsDTO;
import domain.dtos.SwimmerDTO;
import domain.entities.Swimmer;
import domain.enums.SwimmingDistance;
import domain.enums.SwimmingStyle;
import javafx.application.Platform;
import observer.SwimmingRaceObserver;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import protocol.requests.*;
import protocol.responses.*;

import java.io.IOException;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.net.Socket;
import java.net.SocketException;
import java.security.NoSuchAlgorithmException;
import java.util.List;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.LinkedBlockingQueue;

public class SwimmingRaceServicesProxy implements SwimmingRaceServices {

    private final String host;
    private final int port;
    private SwimmingRaceObserver client;
    private ObjectOutputStream outputStream;
    private ObjectInputStream inputStream;
    private Socket socket;
    private final BlockingQueue<Response> qresponses;
    private volatile boolean finished;
    private final static Logger logger = LogManager.getLogger();

    public SwimmingRaceServicesProxy(String host, int port) {
        this.host = host;
        this.port = port;
        qresponses = new LinkedBlockingQueue<>();
    }

    @Override
    public void login(String username, String password, SwimmingRaceObserver client) throws NoSuchAlgorithmException, ServicesException {
        initializeConnection();
        AdminDTO adminDTO = new AdminDTO(username, password);
        logger.info("Sending login request: username: {}, password: {}...", username, password);
        sendRequest(new LoginRequest(adminDTO));
        logger.info("Waiting for response...");
        Response response = readResponse();
        if (response instanceof OkResponse) {
            logger.info("OkResponse received!");
            this.client = client;
        }
        if (response instanceof ErrorResponse errorResponse) {
            logger.info("ErrorResponse received!");
            closeConnection();
            throw new ServicesException(errorResponse.getErrorMessage());
        }
    }

    private void initializeConnection() {
        try {
            logger.info("Connection details: host: {}, port: {}", host, port);
            socket = new Socket(host, port);
            logger.info("Successfully connected!");
            outputStream = new ObjectOutputStream(socket.getOutputStream());
            outputStream.flush();
            inputStream = new ObjectInputStream(socket.getInputStream());
            finished = false;
            startReader();
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    private void startReader() {
        Thread tw = new Thread(new ReaderThread());
        logger.info("Starting response reader thread...");
        tw.start();
    }

    private void closeConnection() {
        logger.info("Closing connection...");
        finished = true;
        try {
            inputStream.close();
            outputStream.close();
            socket.close();
            client = null;
            logger.info("Connection closed successfully!");
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    private void sendRequest(Request request) {
        try {
            outputStream.writeObject(request);
            outputStream.flush();
            logger.info("Request sent successfully!");
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    private Response readResponse() {
        Response response = null;
        try {
            response = qresponses.take();
            logger.info("Response read successfully!");
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
        return response;
    }

    @Override
    public void logout(String username) throws ServicesException {
        logger.info("Sending logout request: username: {}", username);
        sendRequest(new LogoutRequest(username));
        logger.info("Waiting for response...");
        Response response = readResponse();
        if (response instanceof OkResponse) {
            logger.info("OkResponse received!");
            closeConnection();
        }
        if (response instanceof ErrorResponse errorResponse) {
            logger.info("ErrorResponse received!");
            throw new ServicesException(errorResponse.getErrorMessage());
        }
    }

    @Override
    public List<RaceDTO> findAllRacesDetails() {
        logger.info("Sending FindAllRacesDetailsRequest...");
        sendRequest(new FindAllRacesDetailsRequest());
        logger.info("Waiting for response...");
        Response response = readResponse();
        if (response instanceof FindAllRacesDetailsResponse findAllRacesDetailsResponse) {
            logger.info("FindAllRacesDetailsResponse received!");
            return findAllRacesDetailsResponse.getAllRacesDetails();
        }
        else {
            logger.info("Wrong response received!");
            return null;
        }
    }

    @Override
    public List<SwimmerDTO> findAllSwimmersDetailsForRace(SwimmingDistance swimmingDistance, SwimmingStyle swimmingStyle) {
        RaceDetailsDTO raceDetailsDTO = new RaceDetailsDTO(swimmingDistance, swimmingStyle);
        logger.info("Sending FindAllSwimmersDetailsForRaceRequest: raceDetails: {}", raceDetailsDTO);
        sendRequest(new FindAllSwimmersDetailsForRaceRequest(raceDetailsDTO));
        logger.info("Waiting for response...");
        Response response = readResponse();
        if (response instanceof FindAllSwimmersDetailsForRaceResponse findAllSwimmersDetailsForRaceResponse) {
            return findAllSwimmersDetailsForRaceResponse.getAllSwimmersDetailsForRace();
        }
        else {
            logger.info("Wrong response received!");
            return null;
        }
    }

    @Override
    public void addSwimmer(String firstName, String lastName, Integer age, List<RaceDetailsDTO> raceDetailsDTOs) {
        SwimmerDTO swimmerDTO = new SwimmerDTO(new Swimmer(firstName, lastName, age), raceDetailsDTOs);
        logger.info("Sending AddSwimmerRequest: swimmerDetails: {}", swimmerDTO);
        sendRequest(new AddSwimmerRequest(swimmerDTO));
        logger.info("Waiting for response...");
        Response response = readResponse();
        if (response instanceof OkResponse) {
            logger.info("OkResponse received!");
        }
        else {
            logger.info("Wrong response received!");
        }
    }

    private void handleUpdate(UpdateResponse updateResponse) {
        Platform.runLater(() -> {
            if (updateResponse instanceof RacesUpdatedResponse) {
                logger.info("Updating UI...");
                client.racesUpdated();
                logger.info("UI updated successfully!");
            }
        });
    }

    private class ReaderThread implements Runnable {
        @Override
        public void run() {
            while (!finished) {
                try {
                    Object response = inputStream.readObject();
                    if (response instanceof UpdateResponse) {
                        logger.info("UpdateResponse received. Handling response...");
                        handleUpdate((UpdateResponse) response);
                    }
                    else {
                        try {
                            qresponses.put((Response) response);
                        } catch (InterruptedException e) {
                            e.printStackTrace();
                        }
                    }
                }
                catch (SocketException e) {
                    //e.printStackTrace();
                }
                catch (IOException | ClassNotFoundException e) {
                    e.printStackTrace();
                }
            }
        }
    }
}

File: Client/Services/SwimmingRaceServicesProxy.cs
using System;
using System.Collections.Generic;
using System.Net.Sockets;
using System.Runtime.Remoting;
using System.Runtime.Serialization;
using System.Runtime.Serialization.Formatters.Binary;
using System.Threading;
using log4net;
using Model.Domain.DTOs;
using Model.Domain.Entities;
using Model.Domain.Enums;
using Model.Observer;
using Model.Protocol.Requests;
using Model.Protocol.Responses;
using Model.Services;

namespace Client.Services;

public class SwimmingRaceServicesProxy : ISwimmingRaceServices
{
    private static readonly ILog Logger = LogManager.GetLogger("Proxy");
    private readonly string _host;
    private readonly int _port;
    private readonly Queue<IResponse> _qresponses;
    private ISwimmingRaceObserver _client;
    private volatile bool _finished;
    private IFormatter _formatter;
    private NetworkStream _networkStream;
    private TcpClient _socketClient;
    private EventWaitHandle _waitHandle;

    public SwimmingRaceServicesProxy(string host, int port)
    {
        _host = host;
        _port = port;
        _qresponses = new Queue<IResponse>();
    }

    public void Login(string username, string password, ISwimmingRaceObserver client)
    {
        InitializeConnection();
        var adminDTO = new AdminDTO(username, password);
        Logger.InfoFormat("Sending login request: username: {0}, password: {1}...", username, password);
        SendRequest(new LoginRequest(adminDTO));
        Logger.Info("Waiting for response...");
        var response = ReadResponse();
        if (response is OkResponse)
        {
            Logger.Info("OkResponse received!");
            _client = client;
        }

        if (response is ErrorResponse errorResponse)
        {
            Logger.Info("ErrorResponse received!");
            CloseConnection();
            throw new ServicesException(errorResponse.ErrorMessage);
        }
    }

    public void Logout(string username)
    {
        Logger.InfoFormat("Sending logout request: username: {0}", username);
        SendRequest(new LogoutRequest(username));
        Logger.Info("Waiting for response...");
        var response = ReadResponse();
        if (response is OkResponse)
        {
            Logger.Info("Okresponse received!");
            CloseConnection();
        }

        if (response is ErrorResponse errorResponse)
        {
            Logger.Info("ErrorResponse received!");
            throw new ServerException(errorResponse.ErrorMessage);
        }
    }

    public List<RaceDTO> FindAllRacesDetails()
    {
        Logger.Info("Sendinf FindAllRacesDetailsRequest...");
        SendRequest(new FindAllRacesDetailsRequest());
        Logger.Info("Waiting for response...");
        var response = ReadResponse();
        if (response is FindAllRacesDetailsResponse findAllRacesDetailsResponse)
        {
            Logger.Info("FindAllRacesDetailsResponse received!");
            return findAllRacesDetailsResponse.AllRacesDetails;
        }

        Logger.Info("Wrong response received!");
        return null;
    }

    public List<SwimmerDTO> FindAllSwimmersDetailsForRace(SwimmingDistance swimmingDistance,
        SwimmingStyle swimmingStyle)
    {
        var raceDetailsDto = new RaceDetailsDTO(swimmingDistance, swimmingStyle);
        Logger.InfoFormat("Sending FindAllSwimmersDetailsForRaceRequest: raceDetails: {0}", raceDetailsDto);
        SendRequest(new FindAllSwimmersDetailsForRaceRequest(raceDetailsDto));
        Logger.Info("Waiting for response...");
        var response = ReadResponse();
        if (response is FindAllSwimmersDetailsForRaceResponse findAllSwimmersDetailsForRaceResponse)
        {
            return findAllSwimmersDetailsForRaceResponse.AllSwimmersDetailsForRace;
        }

        Logger.Info("Wrong rasponse received!");
        return null;
    }

    public void AddSwimmer(string firstName, string lastName, int age, List<RaceDetailsDTO> raceDetailsDTOs)
    {
        var swimmerDTO = new SwimmerDTO(new Swimmer(firstName, lastName, age), raceDetailsDTOs);
        Logger.InfoFormat("Sending AddSwimmerRequest: swimmerDetails: {0}", swimmerDTO);
        SendRequest(new AddSwimmerRequest(swimmerDTO));
        Logger.Info("Waiting for response...");
        var response = ReadResponse();
        if (response is OkResponse)
            Logger.Info("OkResponse received!");
        else
            Logger.Info("Wrong response received!");
    }

    private void InitializeConnection()
    {
        try
        {
            Logger.InfoFormat("Connection details: host: {0}, port: {1}", _host, _port);
            _socketClient = new TcpClient(_host, _port);
            Logger.Info("Successfully connected!");
            _networkStream = _socketClient.GetStream();
            _formatter = new BinaryFormatter();
            _finished = false;
            _waitHandle = new AutoResetEvent(false);
            StartReader();
        }
        catch (Exception ex)
        {
            Console.WriteLine(ex.StackTrace);
        }
    }

    private void StartReader()
    {
        var t = new Thread(Run);
        t.Start();
    }

    private void HandleUpdate(IUpdateResponse updateResponse)
    {
        if (updateResponse is RacesUpdatedResponse)
        {
            Logger.Info("Updating UI...");
            _client.RacesUpdated();
            Logger.Info("UI updated successfully!");
        }
    }

    private void CloseConnection()
    {
        Logger.Info("Closing connection...");
        _finished = true;
        try
        {
            _networkStream.Close();
            _socketClient.Close();
            _waitHandle.Close();
            _client = null;
            Logger.Info("Connection closed successfully!");
        }
        catch (Exception ex)
        {
            Console.WriteLine(ex.StackTrace);
        }
    }

    private void SendRequest(IRequest request)
    {
        try
        {
            _formatter.Serialize(_networkStream, request);
            _networkStream.Flush();
            Logger.Info("Request sent successfully!");
        }
        catch (Exception ex)
        {
            Console.WriteLine(ex.StackTrace);
        }
    }

    private IResponse ReadResponse()
    {
        IResponse response = null;
        try
        {
            _waitHandle.WaitOne();
            lock (_qresponses)
            {
                response = _qresponses.Dequeue();
                Logger.Info("Response read successfully!");
            }
        }
        catch (Exception ex)
        {
            Console.WriteLine(ex.StackTrace);
        }

        return response;
    }

    public void Run()
    {
        while (!_finished)
            try
            {
                var response = _formatter.Deserialize(_networkStream);
                if (response is IUpdateResponse)
                {
                    Logger.Info("UpdateResponse received. Handling response...");
                    HandleUpdate((IUpdateResponse) response);
                }
                else
                {
                    lock (_qresponses)
                    {
                        _qresponses.Enqueue((IResponse) response);
                    }

                    _waitHandle.Set();
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.StackTrace);
            }
    }
}